voip_utils/sip.py

```diff
@@ -168,6 +175,79 @@ def __init__(self, sdp_info: SdpInfo) -> None:
         """Set up SIP server."""
         self.sdp_info = sdp_info
         self.transport = None
+        self._outgoing_calls: dict[str, int] = {}
+
+    def outgoing_call(
+        self, source: SipEndpoint, destination: SipEndpoint, rtp_port: int
+    ):
+        """Make an outgoing call from the given source endpoint to the destination endpoint, using the rtp_port for the local RTP port of the call."""
+        session_id = str(time.monotonic_ns())
+        session_version = session_id
+        call_id = session_id
+        self._register_outgoing_call(call_id, rtp_port)
+        if self.transport is None:
+            _LOGGER.warning("No transport for outgoing VOIP calls")
+            return
+
+        sdp_lines = [
+            "v=0",
+            f"o={source.username} {session_id} {session_version} IN IP4 {source.host}",
+            "s=Talk",
+            f"c=IN IP4 {source.host}",
+            "t=0 0",
+            f"m=audio {rtp_port} RTP/AVP 123 96 101 103 104",
+            "a=sendrecv",
+            "a=rtpmap:96 opus/48000/2",
+            "a=fmtp:96 useinbandfec=0",
+            "a=rtpmap:123 opus/48000/2",
+            "a=fmtp:123 maxplaybackrate=16000",
+            "a=rtpmap:101 telephone-event/48000",
+            "a=rtpmap:103 telephone-event/16000",
+            "a=rtpmap:104 telephone-event/8000",
+            "a=ptime:20",
+            "",
+        ]
+        sdp_text = _CRLF.join(sdp_lines)
+        sdp_bytes = sdp_text.encode("utf-8")
+
+        invite_lines = [
+            f"INVITE {destination.uri} SIP/2.0",
+            f"Via: SIP/2.0/UDP {source.host}:{source.port}",
+            f"From: {source.sip_header}",
+            f"Contact: {source.sip_header}",
+            f"To: {destination.sip_header}",
+            f"Call-ID: {call_id}",
+            "CSeq: 50 INVITE",
+            "User-Agent: test-agent 1.0",
+            "Allow: INVITE, ACK, OPTIONS, CANCEL, BYE, SUBSCRIBE, NOTIFY, INFO, REFER, UPDATE",
+            "Accept: application/sdp, application/dtmf-relay",
+            "Content-Type: application/sdp",
+            f"Content-Length: {len(sdp_bytes)}",
+            "",
+        ]
+        invite_text = _CRLF.join(invite_lines) + _CRLF
+        invite_bytes = invite_text.encode("utf-8")
+
+        msg_bytes = invite_bytes + sdp_bytes
+
+        _LOGGER.debug(msg_bytes)
+
+        self.transport.sendto(
+            msg_bytes,
+            (destination.host, destination.port),
+        )
+
+    def _register_outgoing_call(self, call_id: str, rtp_port: int):
+        """Register the RTP port associated with an outgoing call."""
+        self._outgoing_calls[call_id] = rtp_port
+
+    def _get_call_rtp_port(self, call_id: str):
```

Comment:
consider adding return values too. 
```suggestion
    def _get_call_rtp_port(self, call_id: str) -> int | None:
```